README
=======
dacapo
=======

Note
====
I'd really appreciate any feedback. 
Send an E-Mail (german or english) to claw <dot> strophob <at> gmx <dot> de or
visit the project-homepage at http://sourceforge.net/projects/dacapo-audio
where you can open a ticket (if you found a bug) or write a review or open open a discussion.
Also, I'd welcome any translator, anybody who would create RPM- or DEB-Packages, 
as well as Windows-Installer.

Introduction
============
First: 
*What is dacapo?* It is a small, "lightweight" music
Player, with a few special features.
*What is dacapo not?* A system for managing music collections.
It can't create playlists, will not do changes in the audio data
nor organize any directories.
*Why another player?* Because most players neither show the images properly
nor prepare the metadata free configurable, nor display the songtext, etc., 
and if they can, they will not run on my little VIA M800 because they can
way too much and eat too much resources.

I developed this Python package to play my digitized Music Collection (FLAC) 
on my small, quiet (fanless) VIA EPIA M800, and I like the metadata ("tags", 
ie information such as "Artist", "Title", "Album", etc.) and the cover 
(also full screen) to display.
The images (cover, back cover, band photos) are stored in my FLAC files. 
Sense and nonsense about this one can argue, it fills forums. 
In any case, for a full screen display has a reasonable quality of the images 
to be added, otherwise it's no fun.
And this does no tool which automatically fetches the images.
How it comes then, that my buddy saw it, loved it, and so the
Program has been expanded by some modules.
In addition, my buddy uses Windows ...

It's just a player, no management of the collection! This is done by
excellent programs such as Quod Libet [2], or MP3Tag [3] in Windows.
With these programs, the metadata can be maintained, as well as fine
Playlists could be created, which are then processed by dacapo.
With MP3Tag or easytag [4] you can easily put the images in the files. 
For the ReplayGain analysis/storage I recommend RGain [5].

To display the images and data pyGame is used because IMHO it is the
thinnest and fastest performing option.
To play the audio I use the GStreamer, because he plays pretty much
everything, and has Replay Gain [1] and gapless support.
Since GStreamer requires Gtk+ anyway, I've used this for the GUI, too.

Features
--------
Thus, dacapo has now the following features:
 - Replay Gain [1] is supported
 - Gapless (gapless playback) is supported
   (important during live music or transitions)
 - All metadata can be displayed, which is stored in the
   Audio files - freely configurable
 - All images (what is stored in the audio files)
   can be shown. e.g. as slideshow
 - Lyrics can be displayed (as an additional image).
 - Synchronized Lyrics can be displayed (like "karaoke")
 - Command line and graphical call possible
 - Runs on Linux and Windows
 - Runs on "small" computers (800 MHz with 1 GB RAM)
 - Is freely configurable
 - Plays (currently) FLAC and MP3, other formats are in the preparation
 - Multilingual (Translator welcome)


Of course it stands and falls with the data. What is not stored in the audio 
files can not be displayed.


[1] http://replaygain.org
[2] http://code.google.com/p/quodlibet
[3] http://www.mp3tag.de/
[4] http://easytag.sourceforge.net/
[5] http://bitbucket.org/fk/rgain

Requirements
============
 - Python 2.6 or 2.7 -- http://python.org/
 - Gtk+ 2.24 -- http://www.gtk.org/
 - PyGTK 2.24 -- http://www.pygtk.org/
 - pyGame 1.9 -- http://www.pygame.org/
 - Mutagen -- http://code.google.com/p/mutagen/
 - GStreamer 0.10 -- http://gstreamer.org/
 - PyGST 0.10 -- http://gstreamer.freedesktop.org/modules/gst-python.html

Before installing: Python, GTK and PyGTK 2.24, GStreamer
and PyGST 0.10 should be installed.
Mutagen and pyGame should be installed automatically.


Installation
============
Just install it like any other Python package: Unpack, then (as root/with
**sudo**)

 # python setup.py install

After installation, the configuration file should be adapted.
This should be located in the user directory:

 # ~/.dacapo/dacapo.conf

Linux example:
  /home/claw/.dacapo/dacapo.conf

Windows example:
  C:\Dokumente und Einstellungen\Claw\.dacapo\dacapo.conf

Should something gone wrong during the installation,
the directory could be copied manually from the package.
Copy the contents of the package directory dacapo/data in the
directory mentioned above.


Program start
=============

The scripts dacapo and dacapoui should have been copied during installation
in the directory */usr/local/bin* or on Windows in the python subdirectory *Scripts*.

**dacapo**
=============
This is the command line part.
Basic usage is simple:

 $ dacapo AUDIOFILE1 AUDIOFILE2 ...
 or
 $ dacapo PLAYLIST 
 or
 $ dacapo /PATH/TO/MY/MUSIC/

There are some options; see them by running

 $ dacapo --help

**dacapoui**
=============
Thre you go with the graphical part (Gtk+).
Simple:

 $ dacapoui

**Function-Keys**
=================

The following keys have a function during operation:
 - HOME = first song of the playlist
 - END = Last song of the playlist
 - SPACE = pause / start
 - LEFT / RIGHT = + / -10 seconds
 - Up / Down = next / previous song
 - ESC / Q = Quit
 - F = Full screen / window

Copyright
=========

Copyright (c) 2013 Thomas Korell <claw.strophob@gmx.de>


CHANGES.txt
V0.1.9i, 2014-11-30 --  Make meta-positioning relative available

V0.1.9i, 2014-11-15 --  Moved project to github

V0.1.9i, 2014-05-13 --  Filter in OpenFileDialog from QtSyncLyrics extended to *.*
						Help for Fonts
						
V0.1.9h, 2014-05-13 --  Bug in dacapoui: Call from Menu has no sys.stdin.encoding
						Bug in dacapo: Keys not accepted after window-mode
							changed

V0.1.9g, 2014-04-23 --  Bug in QtFlac2Mp3, Translations

V0.1.9f, 2014-03-19 --  Textmargin and seconds for fast forward/backward
						taken from config-file
						first part of i18n for QtFlac2Mp3 with gettext()
V0.1.9e, 2014-02-08 -- 	Changing QtFlac2Mp3 to take parameters from 
						the commandline
V0.1.9d, 2014-02-01 -- 	Adding QtSyncLyrics and QtFlac2Mp3
V0.1.9c, 2014-01-01 -- 	configuring the time-display after title causes
						an error. (ticket #9).
V0.1.9b, 2013-11-17 -- 	Versioncontrol, changing setup.py
V0.1.9a, 2013-09-23 -- 	Adding manpages, changing setup.py

V0.1.9, 2013-09-23 -- 	Windows: supporting filenames with special characters
						(internal playlist only follows UTF-8)
						some new features in dacapoGST.py for LRC-generator QtSyncLyrics
                        

						
V0.1.2c, 2013-07-13 -- 	Remove TK from errorhandling, replace it by Gtk+
						Check, if the path to the logfile is existing
						Create this CHANGES.txt
						Some documentation in the modules.
V0.1.3, 2013-07-26 -- 	actual title in gapless-mode necessary for synced-lyrics
						starting FAQ
						Remove TK from createconfig, replace it by Gtk+
V0.1.3a, 2013-08-01 -- 	Last line of synced lyrics was flickering. Fixed.
						Two new points in FAQ
V0.1.4, 2013-08-04 -- 	Changing the name of the project to "dacapo" due to name conflicts.
						A new point in FAQ
V0.1.5, 2013-08-04 -- 	Setup -> existing config-dir: Bug in Gtk-Question/return
						Creating modules "vorbis.py" and "mp3.py" for FLAC- and MP3-metadata
						as an subclass from audiofile
						Adding Ogg-Vorbis
						Bug from renaming gplayer -> dacapo: dacapoui cannot import dacapo
V0.1.6, 2013-08-12 -- 	Renaming Module metadaten -> metadata
						Adding mimehelp to metadata
						Adding WMA-format (ticket 1)
						Bugfix: getMetaData() returned "  " if key was not found
						Bugfix: Windows -> import gtk failed  (ticket 4)
						Bugfix: Windows -> setup-overwrite-config-message 
#								OK-Button=win32con.IDOK (ticket 3)					
						Bugfix: MP3 -> additional Pictures
V0.1.7, 2013-08-16 -- 	requirements: setuptools
						setup.py: adding classifiers
						Bugfix: center pictures in Diamode
						Bugfix: Switching between window and fullscreen - 
							unnecessary cleaning screen  (ticket 6)
						Bugfix: synced lyrics from lrc-file with spacelines  (ticket 5)
V0.1.8, 2013-08-28 -- 	use empty LRC-lines to display blank line
                        merge dacapo.conf and the folder instead of replacing   (ticket 7)
                        create dacapo.conf.sample.eng and dacapo.conf.sample.ger as 
                        documentation for dacapo.conf
                        new option "showLyricsWhenSynced" 
                        new option "lyricFontSpace" 
                        Bugfix: GStreamer overwrites cmdline-help  (ticket 8)


ez_setup.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# from distutils.core import setup
from distutils.command.install import install
from setuptools import setup, find_packages
import sys, os

with open('README.txt') as file:
    long_description = file.read()

print("Setup-Mode: %s" % (sys.argv[1]) )
if sys.argv[1] == "sdist" or \
	sys.argv[1] == "bdist" :    
	import shutil
	src_file = "dacapo/data/VERSION" 
	dest_file = "./VERSION" 
	shutil.copy(src_file, dest_file)

VERSION = open("VERSION").read().strip()

class dacapo_install(install):
	description = "Custom Install Process"
	user_options= install.user_options[:]
	user_options.extend([('manprefix=', None, 
		'MAN Prefix Path if not /usr/local/share/>')])

	def initialize_options(self):
		self.manprefix = None
		install.initialize_options(self)

	def finalize_options(self):
		if self.manprefix is None :
			self.manprefix = "/usr/local/share/"
		install.finalize_options(self)

	def install_manpages(self):
		import platform, string, shutil
		if platform.system() <> 'Windows':
			print "copy manpages..."
			for root, dirs, files in os.walk('./man/'):
				for filename in files:
					src_file = os.path.join(root, filename) 
					man_root = os.path.join(self.manprefix, \
						string.lstrip(root, './'))
					dest_file = os.path.join(man_root, filename)
					if not os.path.exists(man_root):
						os.makedirs(man_root) 					
					print("copy %s -> %s " % (src_file, dest_file))
					shutil.copy(src_file, dest_file)
			
	def run(self):
		install.run(self)
		# Custom stuff here
		# distutils.command.install actually has some nice helper methods
		# and interfaces. I strongly suggest reading the docstrings.

		if sys.argv[1] == "install" :
			print "installiere config"
			import dacapo.config.createconfig
			self.install_manpages()
		

setup(
    name = "dacapo",
    version=VERSION,
    packages = ['dacapo', 'dacapo.ui', 'dacapo.config', 'dacapo.data', 
    'dacapo.errorhandling', 'dacapo.metadata', 'dacapo.playlist'],
    scripts = ["bin/dacapo", "bin/dacapoui"],

    # Project uses reStructuredText, so ensure that the docutils get
    # installed or upgraded on the target machine
    requires=[
			'pygame (>=1.9)',
			'argparse (>=1.1)',
			'setuptools (>=0.9)',
			'mutagen (>=1.21)'
			],
    install_requires = [
			# 'gtk==2.24',
			# 'pygtk==2.24', 
			# 'gst>=0-10', 
			# 'pygst>=0.10', 
			'pygame>=1.9',
			'argparse>=1.1',
			'setuptools>=0.9',
			'mutagen>=1.21'
		],

    package_data = {
        # If any package contains *.txt or *.rst files, include them:
        '': ['*.txt', '*.rst'],
        # And include any *.glade files found in the 'dacapo' package, too:
        'dacapo': ['*.glade'],
        # And include any *.tar.gz from the package 'dacapo.config, too:
        'dacapo.config': ['*.gz'],
        'dacapo.data': ['*'],
    },

	# data_files = ['docs'],

    # metadata for upload to PyPI
    author = "Thomas Korell",
    author_email = "claw.strophob@gmx.de",
    description = "Lightweight Music Player with cover- and lyrics-display",
	long_description=long_description,
    license = "GNU General Public License (v2 or later)",
    keywords = "FLAC MP3 Player Coverart lyrics karaoke",
    url = "http://sourceforge.net/projects/dacapo-player",   # project home page, if any

    # could also include long_description, download_url, classifiers, etc.
	# cmdclass={'setconfig': my_install},
	classifiers=[
          'Development Status :: 4 - Beta',
          'Environment :: Win32 (MS Windows)',
          'Environment :: X11 Applications',
          'Intended Audience :: End Users/Desktop',
          'Intended Audience :: Developers',
          'License :: OSI Approved :: GNU General Public License v2 or later (GPLv2+)',
          'Operating System :: Microsoft :: Windows',
          'Operating System :: POSIX',
          'Programming Language :: Python',
          'Topic :: Multimedia :: Sound/Audio :: Players',
          'Topic :: Multimedia :: Sound/Audio :: Players :: MP3',
          ],
	cmdclass={"install": dacapo_install},

)

